subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/ClientProvidedPhasedActionRunner.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.tooling.internal.provider.runner;
+
+import org.gradle.BuildAdapter;
+import org.gradle.BuildResult;
+import org.gradle.api.BuildCancelledException;
+import org.gradle.api.internal.GradleInternal;
+import org.gradle.api.invocation.Gradle;
+import org.gradle.initialization.BuildEventConsumer;
+import org.gradle.internal.invocation.BuildAction;
+import org.gradle.internal.invocation.BuildActionRunner;
+import org.gradle.internal.invocation.BuildController;
+import org.gradle.tooling.internal.protocol.InternalBuildActionVersion2;
+import org.gradle.tooling.internal.protocol.InternalBuildActionFailureException;
+import org.gradle.tooling.internal.protocol.InternalBuildCancelledException;
+import org.gradle.tooling.internal.protocol.InternalPhasedAction;
+import org.gradle.tooling.internal.provider.BuildActionResult;
+import org.gradle.tooling.internal.provider.ClientProvidedPhasedAction;
+import org.gradle.tooling.internal.provider.PhasedBuildActionResult;
+import org.gradle.tooling.internal.provider.serialization.PayloadSerializer;
+
+import javax.annotation.Nullable;
+
+public class ClientProvidedPhasedActionRunner implements BuildActionRunner {
+    @Override
+    public void run(BuildAction action, final BuildController buildController) {
+        if (!(action instanceof ClientProvidedPhasedAction)) {
+            return;
+        }
+
+        GradleInternal gradle = buildController.getGradle();
+
+        gradle.getStartParameter().setConfigureOnDemand(false);
+
+        ClientProvidedPhasedAction clientProvidedPhasedAction = (ClientProvidedPhasedAction) action;
+        PayloadSerializer payloadSerializer = getPayloadSerializer(gradle);
+
+        InternalPhasedAction phasedAction = (InternalPhasedAction) payloadSerializer.deserialize(clientProvidedPhasedAction.getPhasedAction());
+
+        addBuildListener(phasedAction, buildController);
+
+        // We don't know if the model builders invoked in the after configuration action will add tasks to the graph or not
```

Comment:
This should be improved in `buildController` by providing an API similar to

```
buildController.build {
  loadSettings()
  configureBuild()
  runTasks()
}
```

That method would automatically fire the right buildStarted/buildFinished hooks and you would get a failure if you try to do things in the wrong order. Using such an API you could run the build only up to the point that you want and react to what happened in between.

This can be a follow-up improvement, but until this is done the `projectsLoaded` hook is kinda unhelpful, because it would still configure everything even though the user only wanted information based on the settings file.

Will you have time to look into this after we merge this change?

Reply:
Thank you for the suggestion :) I was thinking about a solution for that.
I'll try to implement it in another commit in this same pull request soon, so it doesn't block review for now and hopefully we can merge together.

Reply:
@oehme, I looked into this and implemented a solution based on this suggestion. As it is kind of orthogonal to this change, I thought it was better to submit as a separate PR (it touches progress events to make Loading projects and Configure Projects two separate build operations).

The commit is [here](https://github.com/lsmaira/gradle/tree/build-controller-api) and once we merge this PR, I'll rebase on top of master and send another PR.

Is it okay?